autopilot/contracts.go

```diff
@@ -1,27 +1,417 @@
+		// grab contract metadata
+		metadata := contract.Metadata
+
+		// update metadata:
+		// TODO: is host in host DB
+		// TODO: is max revision
+		// TODO: is offline
+		// TODO: is gouging
+		// TODO: is low score
+		// TODO: is out of funds
+
+		// TODO: is up for renewal
+		// TODO: is GFU limited (allowance # hsots)
+
+		// update contract metadata
+		err = ap.bus.UpdateContractMetadata(contract.ID, metadata)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ap *Autopilot) runContractRenewals(cfg Config, s State, budget *types.Currency, renterAddress types.UnlockHash) ([]worker.Contract, error) {
+	// fetch all contracts that are up for renew
+	toRenew, err := ap.renewableContracts(s.BlockHeight + cfg.Contracts.RenewWindow)
+	if err != nil {
+		return nil, err
+	}
+
+	// perform the renewals
+	var renewed []worker.Contract
+	for _, renew := range toRenew {
+		// check our budget
+		renterFunds, err := ap.renewFundingEstimate(cfg, s, renew.ID)
+		if budget.Cmp(renterFunds) < 0 {
+			break
+		}
+
+		// derive the renter key
+		renterKey := ap.deriveRenterKey(renew.HostKey)
+		if err != nil {
+			return nil, err
+		}
+
+		var hostCollateral types.Currency // TODO
+		contract, err := ap.renewContract(cfg, s, renew, renterKey, renterAddress, renterFunds, hostCollateral)
+		if err != nil {
+			// TODO: handle error properly, if the wallet ran out of outputs
+			// here there's no point in renewing more contracts until a
+			// block is mined, maybe we could/should wait for pending transactions?
+			return nil, err
+		}
+
+		// update the budget
+		*budget = budget.Sub(renterFunds)
+
+		// persist the contract
+		err = ap.bus.AddContract(contract)
+		if err != nil {
+			return nil, err
+		}
+
+		// persist the metadata
+		err = ap.bus.UpdateContractMetadata(contract.ID(), bus.ContractMetadata{
+			RenewedFrom: renew.ID,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// add to set
+		renewed = append(renewed, worker.Contract{
+			HostKey:   renew.HostKey,
+			HostIP:    renew.HostIP,
+			ID:        contract.ID(),
+			RenterKey: renterKey,
+		})
+	}
+
+	return renewed, nil
+}
+
+func (ap *Autopilot) runContractFormations(cfg Config, s State, budget *types.Currency, renterAddress types.UnlockHash) ([]worker.Contract, error) {
+	// fetch all active contracts
+	active, err := ap.bus.ActiveContracts(math.MaxUint64)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch recommended txn fee
+	fee, err := ap.bus.RecommendedFee()
+	if err != nil {
+		return nil, err
+	}
+
+	// calculate min/max contract funds
+	allowance := cfg.Contracts.Allowance.Div64(cfg.Contracts.Hosts)
+	maxInitialContractFunds := allowance.Div64(10) // TODO: arbitrary divisor
+	minInitialContractFunds := allowance.Div64(20) // TODO: arbitrary divisor
+
+	// form missing contracts
+	var formed []worker.Contract
+	missing := int(cfg.Contracts.Hosts) - len(active) // TODO: add leeway so we don't form hosts if we dip slightly under `needed` (?)
```

Comment:
Some leeway is probably good. But it should only be applied when we have at least `cfg.Contracts.Hosts` contracts. e.g. we don't form new contracts if 40 out of 50 are active but if 40 out of 40 are active and we want 50 we still form 10 more.

Reply:
👍  ok - will take this into account when I add the leeway